File: Python_examples/quick_sort.py
import random as r

array = [r.randint(1, 50) for i in range(40)]
print(array)


def quick_sort(array, start, end):
    left = start
    right = end
    pivot = array[int(left + (right - left) / 2)]
    while left <= right:
        while array[left] < pivot:
            left += 1
        while array[right] > pivot:
            right -= 1
        if left <= right:
            array[left], array[right] = array[right], array[left]
            left += 1
            right -= 1

    if right > start:
        quick_sort(array, start, right)
    if left < end:
        quick_sort(array, left, end)


quick_sort(array, 0, len(array) - 1)

print(array)

File: C++_examples/quick_sort.cpp
//
// Created by Андрей Кочетков on 03.02.17.
//

void quick_sort(int *a, int n) {
    int i = 0, j = n;        // поставить указатели на исходные места
    int temp, pivot;

    pivot = a[n >> 1];        // центральный элемент

    // процедура разделения
    do {
        while (a[i] < pivot) i++;
        while (a[j] > pivot) j--;

        if (i <= j) {
            temp = a[i];
            a[i] = a[j];
            a[j] = temp;
            i++;
            j--;
        }
    } while (i <= j);


    // рекурсивные вызовы, если есть, что сортировать
    if (j > 0) quick_sort(a, j);

    if (n > i) quick_sort(a + i, n - i);
}

